connect/src/com/telenor/connect/ConnectSdk.java

```diff
@@ -409,8 +457,34 @@ public static IdToken getIdToken() {
      */
     public static void getUserInfo(Callback<UserInfo> userInfoCallback) {
         Validator.sdkInitialized();
-        sConnectIdService.getUserInfo(userInfoCallback);
+        sdkProfile.getConnectIdService().getUserInfo(userInfoCallback);
     }
 
+    /**
+     * MobileConnect stuff
+     */
+
+    public static synchronized void sdkInitializeMobileConnect(
+            Context context,
+            OperatorDiscoveryConfig operatorDiscoveryConfig) {
+        if (isInitialized()) {
+            return;
+        }
+        sdkProfile = new MobileConnectSdkProfile(
+                context,
+                operatorDiscoveryConfig,
+                fetchBooleanProperty(getApplicationInfo(context), USE_STAGING_PROPERTY),
+                fetchBooleanProperty(getApplicationInfo(context), CONFIDENTIAL_CLIENT_PROPERTY));
+        context.registerReceiver(
```

Comment:
Two things here:
1. With this setup you are always listening to sms being received, not just during a sign in process. I think unless this is needed, we should only listen to sms in the window where we actually must. So remember to unregister after you've gotten the SMS that you're looking for, and only listen for it when you are waiting for it.
2. `android.permission.READ_SMS` is a _dangerous_ permission, so you are not automatically granted it and must request it. It seems the code is registering a receiver before requesting this permission. And it seems it's only requesting it on button press, not if the user uses the `authenticate` method. That is not quite consistent with the existing code.

`ConnectWebViewClient` is a bit heavy class, but I would look at that to see how the asynchronous nature of sms is currently handled. The reason for the `Instruction` class is to notify Android to listen for SMS on a given page. If we don't currently have permission to read sms the instruction is stored, then permission is asked for, then the code reads the most recently received sms and starts listening for new ones.